File: web/src/sim/military/Movement.ts
import type { ProvinceGraph } from '~/sim/data/WorldFile'
import { addF32, divF32, f32, mulF32, subF32 } from '~/sim/determinism/float32'
import type { ShortageSystem } from '~/sim/economy/ShortageSystem'
import type { Army } from '~/sim/military/Army'
import { hoursFor } from '~/sim/military/MovementCost'
import { isWater } from '~/sim/world/Terrain'
import type { WorldState } from '~/sim/world/WorldState'

export class MovementOrder {
  #step = 0
  #hoursIntoStep = 0

  constructor(
    readonly armyId: number,
    readonly path: readonly number[],
  ) {}

  get step(): number {
    return this.#step
  }

  get hoursIntoStep(): number {
    return this.#hoursIntoStep
  }

  get isComplete(): boolean {
    return this.#step >= this.path.length - 1
  }

  get nextProvince(): number {
    return this.path[Math.min(this.#step + 1, this.path.length - 1)]!
  }

  advance(hours: number): void {
    this.#hoursIntoStep = addF32(this.#hoursIntoStep, hours)
  }

  completeStep(): void {
    this.#step++
    this.#hoursIntoStep = 0
  }

  hoursRemainingInStep(stepHours: number): number {
    return Math.max(0, subF32(stepHours, this.#hoursIntoStep))
  }
}

/*
 * Menggerakkan pasukan satu jam permainan sekali. Gerakan diselesaikan di tick
 * simulasi, tidak pernah di panggilan balik bingkai, sehingga simpanan yang
 * dimuat di tengah pawai melanjut di titik yang sama berapa pun laju
 * bingkainya.
 */
export class MovementSystem {
  readonly #orders = new Map<number, MovementOrder>()

  /* Boleh kosong supaya gerak tetap bisa diuji sendirian; kalau dipasang,
     inilah yang membuat kelangkaan bahan bakar menelantarkan kolom lapis baja
     di tengah jalan ke garis depan. */
  shortage: ShortageSystem | null = null

  constructor(
    private readonly world: WorldState,
    private readonly land: ProvinceGraph | null = null,
  ) {}

  get pendingOrders(): number {
    return this.#orders.size
  }

  order(army: Army, path: readonly number[]): void {
    if (path.length < 2) {
      this.#orders.delete(army.id)
      return
    }

    if (path[0] !== army.province) {
      throw new RangeError(
        `jalur mulai di ${path[0]} sedangkan pasukan ${army.id} ada di ${army.province}`,
      )
    }

    this.#orders.set(army.id, new MovementOrder(army.id, path))
  }

  cancel(armyId: number): void {
    this.#orders.delete(armyId)
  }

  isMoving(armyId: number): boolean {
    return this.#orders.has(armyId)
  }

  tick(armies: ReadonlyMap<number, Army>, hours = 1): void {
    const finished: number[] = []

    for (const [armyId, order] of this.#orders) {
      const army = armies.get(armyId)
      if (army === undefined || army.isDestroyed) {
        finished.push(armyId)
        continue
      }

      let budget = hours
      while (budget > 0 && !order.isComplete) {
        const stepHours = this.#stepHours(army, order.nextProvince)
        const remaining = order.hoursRemainingInStep(stepHours)

        if (budget < remaining) {
          order.advance(budget)
          budget = 0
          continue
        }

        budget = subF32(budget, remaining)
        const from = army.province
        order.completeStep()
        army.province = order.path[order.step]!

        if (this.#crossedWater(from, army.province)) {
          army.landedOnTick = this.world.clock.tick
        }
      }

      if (order.isComplete) finished.push(armyId)
    }

    for (const armyId of finished) this.#orders.delete(armyId)
  }

  #stepHours(army: Army, destination: number): number {
    const terrain = this.world.provinces.at(destination).terrain
    const speed = Math.max(
      mulF32(army.speed, this.shortage?.speedMultiplierFor(army) ?? 1),
      f32(0.1),
    )
    const bySea = isWater(terrain) || this.#crossedWater(army.province, destination)

    return divF32(hoursFor(terrain, bySea), speed)
  }

  /* Dua provinsi yang hanya tersambung lewat graf laut menuntut naik kapal.
     Tanpa graf darat untuk dibandingkan, setiap langkah diperlakukan sebagai
     jalur darat, yang merupakan pembacaan aman alih-alih pendaratan gratis yang
     diam-diam. */
  #crossedWater(from: number, to: number): boolean {
    if (this.land === null || from === to) return false

    for (const neighbour of this.land.neighboursOf(from)) {
      if (neighbour === to) return false
    }

    return true
  }
}

File: sim/NationRise.Core/Military/Movement.cs
using NationRise.Core.Data;
using NationRise.Core.Economy;
using NationRise.Core.World;

namespace NationRise.Core.Military;

public sealed class MovementOrder
{
    public required int ArmyId { get; init; }
    public required IReadOnlyList<int> Path { get; init; }

    public int Step { get; private set; }
    public float HoursIntoStep { get; private set; }

    public bool IsComplete => Step >= Path.Count - 1;
    public int NextProvince => Path[Math.Min(Step + 1, Path.Count - 1)];

    public void Advance(float hours) => HoursIntoStep += hours;

    public void CompleteStep()
    {
        Step++;
        HoursIntoStep = 0f;
    }

    public float HoursRemainingInStep(float stepHours) => MathF.Max(0f, stepHours - HoursIntoStep);
}

/*
   Moves armies one game hour at a time. Movement is resolved in the simulation
   tick, never in a frame callback, so a save reloaded mid-march resumes at the
   same point regardless of frame rate.
*/
public sealed class MovementSystem(WorldState world, ProvinceGraph? land = null)
{
    private readonly Dictionary<int, MovementOrder> _orders = [];

    /* Optional so movement stays testable on its own; attached, it is what
       makes a fuel shortage strand an armoured column halfway to the front. */
    public ShortageSystem? Shortage { get; set; }

    public int PendingOrders => _orders.Count;

    public void Order(Army army, IReadOnlyList<int> path)
    {
        ArgumentNullException.ThrowIfNull(army);
        ArgumentNullException.ThrowIfNull(path);

        if (path.Count < 2)
        {
            _orders.Remove(army.Id);
            return;
        }

        if (path[0] != army.Province)
        {
            throw new ArgumentException(
                $"Path starts at {path[0]} but army {army.Id} is in {army.Province}.", nameof(path));
        }

        _orders[army.Id] = new MovementOrder { ArmyId = army.Id, Path = path };
    }

    public void Cancel(int armyId) => _orders.Remove(armyId);

    public bool IsMoving(int armyId) => _orders.ContainsKey(armyId);

    public void Tick(IReadOnlyDictionary<int, Army> armies, float hours = 1f)
    {
        var finished = new List<int>();

        foreach ((int armyId, MovementOrder order) in _orders)
        {
            if (!armies.TryGetValue(armyId, out Army? army) || army.IsDestroyed)
            {
                finished.Add(armyId);
                continue;
            }

            float budget = hours;
            while (budget > 0f && !order.IsComplete)
            {
                float stepHours = StepHours(army, order.NextProvince);
                float remaining = order.HoursRemainingInStep(stepHours);

                if (budget < remaining)
                {
                    order.Advance(budget);
                    budget = 0f;
                    continue;
                }

                budget -= remaining;
                int from = army.Province;
                order.CompleteStep();
                army.Province = order.Path[order.Step];

                if (CrossedWater(from, army.Province))
                {
                    army.LandedOnTick = world.Clock.Tick;
                }
            }

            if (order.IsComplete)
            {
                finished.Add(armyId);
            }
        }

        foreach (int armyId in finished)
        {
            _orders.Remove(armyId);
        }
    }

    private float StepHours(Army army, int destination)
    {
        Terrain terrain = world.Provinces[destination].Terrain;
        float speed = MathF.Max(army.Speed * (Shortage?.SpeedMultiplierFor(army) ?? 1f), 0.1f);
        bool bySea = terrain.IsWater() || CrossedWater(army.Province, destination);
        return MovementCost.HoursFor(terrain, bySea) / speed;
    }

    /* Two provinces joined only through the sea graph require embarking. Without
       the land graph to compare against, every step is treated as overland,
       which is the safe reading rather than a silent free landing. */
    private bool CrossedWater(int from, int to)
    {
        if (land is null || from == to)
        {
            return false;
        }

        foreach (ushort neighbour in land.NeighboursOf(from))
        {
            if (neighbour == to)
            {
                return false;
            }
        }

        return true;
    }
}
